src/main/java/rx/observers/TestSubscriber.java

```diff
@@ -271,17 +318,74 @@ public void requestMore(long n) {
      *          if the sequence of items observed does not exactly match {@code items}
      */
     public void assertReceivedOnNext(List<T> items) {
-        testObserver.assertReceivedOnNext(items);
+        if (values.size() != items.size()) {
+            assertionError("Number of items does not match. Provided: " + items.size() + "  Actual: " + values.size()
+            + ".\n"
+            + "Provided values: " + items
+            + "\n"
+            + "Actual values: " + values
+            + "\n");
+        }
+
+        for (int i = 0; i < items.size(); i++) {
+            T expected = items.get(i);
+            T actual = values.get(i);
+            if (expected == null) {
+                // check for null equality
+                if (actual != null) {
+                    assertionError("Value at index: " + i + " expected to be [null] but was: [" + actual + "]\n");
+                }
+            } else if (!expected.equals(actual)) {
+                assertionError("Value at index: " + i 
+                        + " expected to be [" + expected + "] (" + expected.getClass().getSimpleName() 
+                        + ") but was: [" + actual + "] (" + (actual != null ? actual.getClass().getSimpleName() : "null") + ")\n");
+
+            }
+        }
     }
 
     /**
+     * Wait until the current committed value count is less than the expected amount
+     * by sleeping 1 unit at most timeout times and return true if at least
+     * the required amount of onNext values have been received.
+     * @param expected the expected number of onNext events
+     * @param timeout the time to wait for the events
+     * @param unit the time unit of waiting
+     * @return true if the expected number of onNext events happened
+     * @throws InterruptedException if the sleep is interrupted
+     * @since (if this graduates from Experimental/Beta to supported, replace this parenthetical with the release number)
+     */
+    @Experimental
+    public final boolean awaitValueCount(int expected, long timeout, TimeUnit unit) throws InterruptedException {
+        while (timeout != 0 && valueCount < expected) {
+            unit.sleep(1);
```

Comment:
I don't agree. If you wake up with the correct number of values but
slightly late that shouldn't warrant being considered a timeout.

On Thu, Jun 16, 2016 at 1:24 PM Artem Zinnatullin <notifications@github.com>
wrote:

> In src/main/java/rx/observers/TestSubscriber.java
> <https://github.com/ReactiveX/RxJava/pull/4011#discussion_r67387550>:
>
> >      }
> >
> >      /**
> > +     * Wait until the current committed value count is less than the expected amount
> > +     * by sleeping 1 unit at most timeout times and return true if at least
> > +     * the required amount of onNext values have been received.
> > +     * @param expected the expected number of onNext events
> > +     * @param timeout the time to wait for the events
> > +     * @param unit the time unit of waiting
> > +     * @return true if the expected number of onNext events happened
> > +     * @throws InterruptedException if the sleep is interrupted
> > +     * @since (if this graduates from Experimental/Beta to supported, replace this parenthetical with the release number)
> > +     */
> > +    @Experimental
> > +    public final boolean awaitValueCount(int expected, long timeout, TimeUnit unit) throws InterruptedException {
> > +        while (timeout != 0 && valueCount < expected) {
> > +            unit.sleep(1);
>
> Another thing is that sleep(1) does not guarantee that it'll wake you up
> back in 1ms, it may wake you later, so you need to explicitly check that
> expected timeout is not expired: System.nanoTime() - start < timeoutNano
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ReactiveX/RxJava/pull/4011/files/e273a739037cb9c3c935b54c654f1c7321321b91#r67387550>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAEEEURMK6RcFoKC_tRVzW-8G3SLVrSyks5qMYbXgaJpZM4I3eci>
> .
>